File: nowcoder/C++/BM2.链表内指定区间反转.cc
struct ListNode {
	int val;
	struct ListNode *next;
    ListNode(int x) : val(x) {}
};

class Solution {
public:
    /**
     * 
     * @param head ListNode类 
     * @param m int整型 
     * @param n int整型 
     * @return ListNode类
     */
    ListNode* reverseBetween(ListNode* head, int m, int n) {
        ListNode* dummy = new ListNode(0);
        dummy->next = head;
        ListNode* pre = dummy;
        for (int i = 0; i < m-1; ++i) {
            pre = pre->next;
        }
        ListNode* curr = pre->next;
        ListNode* lat = curr->next;
        for (int i = m; i < n; ++i) {
            curr->next = lat->next;
            lat->next = pre->next;
            pre->next = lat;
            lat = curr->next;
        }
        return dummy->next;
    }
};

File: nowcoder/Golang/BM2.链表内指定区间反转.go
package main

// import . "nc_tools"

type ListNode struct {
	Val  int
	Next *ListNode
}

/**
 *
 * @param head ListNode类
 * @param m int整型
 * @param n int整型
 * @return ListNode类
 */
func reverseBetween(head *ListNode, m int, n int) *ListNode {
	var dummy *ListNode = &ListNode{Val: 0, Next: head}
	var pre *ListNode = dummy
	for i := 0; i < m-1; i++ {
		pre = pre.Next
	}
	var curr = pre.Next
	var lat = curr.Next
	for i := m; i < n; i++ {
		curr.Next = lat.Next
		lat.Next = pre.Next
		pre.Next = lat
		lat = curr.Next
	}
	return dummy.Next
}
